tarpc/src/generic_transport/mod.rs

```diff
@@ -0,0 +1,407 @@
+        Transport {
+            inner: FramedRead::new(
+                FramedWrite::new(Framed::new(inner, LengthDelimitedCodec::new()), encoder),
+                decoder,
+            ),
+        }
+    }
+}
+
+#[cfg(feature = "tcp")]
+#[cfg_attr(docsrs, doc(cfg(feature = "tcp")))]
+/// TCP support for generic transport using Tokio.
+pub mod tcp {
+    use {
+        super::*,
+        futures::ready,
+        std::{marker::PhantomData, net::SocketAddr},
+        tokio::net::{TcpListener, TcpStream, ToSocketAddrs},
+    };
+
+    mod private {
+        use super::*;
+
+        pub trait Sealed {}
+
+        impl<Item, ItemDecoder, SinkItem, SinkItemEncoder> Sealed
+            for Transport<TcpStream, Item, ItemDecoder, SinkItem, SinkItemEncoder>
+        {
+        }
+    }
+
+    /// TCP extensions for the generic transport.
+    pub trait TransportExt: private::Sealed {
+        /// Returns the peer address of the underlying TcpStream.
+        fn peer_addr(&self) -> io::Result<SocketAddr>;
+        /// Returns the local address of the underlying TcpStream.
+        fn local_addr(&self) -> io::Result<SocketAddr>;
+    }
+
+    impl<Item, ItemDecoder, SinkItem, SinkItemEncoder> TransportExt
+        for Transport<TcpStream, Item, ItemDecoder, SinkItem, SinkItemEncoder>
+    {
+        fn peer_addr(&self) -> io::Result<SocketAddr> {
+            self.inner.get_ref().get_ref().get_ref().peer_addr()
+        }
+        fn local_addr(&self) -> io::Result<SocketAddr> {
+            self.inner.get_ref().get_ref().get_ref().local_addr()
+        }
+    }
+
+    /// Returns a new JSON transport that reads from and writes to `io`.
+    pub fn new<Item, ItemDecoder, SinkItem, SinkItemEncoder>(
+        io: TcpStream,
+        decoder: ItemDecoder,
+        encoder: SinkItemEncoder,
+    ) -> Transport<TcpStream, Item, ItemDecoder, SinkItem, SinkItemEncoder>
+    where
+        Item: for<'de> Deserialize<'de>,
+        ItemDecoder: Deserializer<Item>,
+        SinkItem: Serialize,
+        SinkItemEncoder: Serializer<SinkItem>,
+    {
+        Transport::from((io, decoder, encoder))
+    }
+
+    /// Connects to `addr`, wrapping the connection in a JSON transport.
+    pub async fn connect<
+        A,
+        Item,
+        ItemDecoder,
+        ItemDecoderFn,
+        SinkItem,
+        SinkItemEncoder,
+        SinkItemEncoderFn,
+    >(
+        addr: A,
+        (decoder_fn, encoder_fn): (ItemDecoderFn, SinkItemEncoderFn),
+    ) -> io::Result<Transport<TcpStream, Item, ItemDecoder, SinkItem, SinkItemEncoder>>
+    where
+        A: ToSocketAddrs,
+        Item: for<'de> Deserialize<'de>,
+        ItemDecoder: Deserializer<Item>,
+        ItemDecoderFn: FnOnce() -> ItemDecoder,
+        SinkItem: Serialize,
+        SinkItemEncoder: Serializer<SinkItem>,
+        SinkItemEncoderFn: FnOnce() -> SinkItemEncoder,
+    {
+        Ok(new(
+            TcpStream::connect(addr).await?,
+            (decoder_fn)(),
```

Comment:
What's the benefit of having the arg be a FnOnce rather than the actual decoder or encoder?